### ISSUE TITLE
Unexpected error at contrib.seq2seq's BeamSearchDecoder

### ISSUE THREAD

**Issue by shuuki4** [tensorflow/tensorflow#9855]
### System information
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: MacOS
- **TensorFlow installed from (source or binary)**: Source
- **TensorFlow version (use command below)**:  v1.1.0-rc2-773-g7fa0cf39f
- **Bazel version (if compiling from source)**:  0.4.5
- **CUDA/cuDNN version**:  None
- **GPU model and memory**: CPU

### Describe the problem

I encountered an unexpected error at tf.contrib.seq2seq's BeamSearchDecoder, *when the beam width is smaller than number of vocabs*.
This is a part of `_beam_search_step` operation in `beam_search_decoder.py`:

```
  scores = _get_scores(
      log_probs=total_probs,
      sequence_lengths=new_prediction_lengths,
      length_penalty_weight=length_penalty_weight)

  time = ops.convert_to_tensor(time, name="time")
  # During the first time step we only consider the initial beam
  scores_flat = control_flow_ops.cond(
      time > 0,
      lambda: array_ops.reshape(scores, [batch_size, -1]),
      lambda: scores[:, 0])

  # Pick the next beams according to the specified successors function
  next_beam_scores, word_indices = nn_ops.top_k(scores_flat, k=beam_width)
  next_beam_scores.set_shape([static_batch_size, beam_width])
  word_indices.set_shape([static_batch_size, beam_width])
```

Since the shape of `scores` is `[batch_size, beam_width, vocab_size]`,  the shape of `scores_flat` is`[batch_size, vocab_size]` at time step 0. However, if `k < shape[-1]` in `nn.top_k` operation, it just throws `InvalidArgumentError: input must have at least k columns`. Thus the code just throws an error and dies.
I think code should be modified to handle cases where `vocab_size ** n < beam_width`, or at least throw an appropriate error message when the input is `vocab_size < beam_width`.

**Comment by girving** [comment]
Nice catch!  Seems like the `top_k` operation should be shielded with a `min`.  Would be up for submitting a small PR?

**Comment by shuuki4** [comment]
OK, I fixed the issue and made PR #9875

**Comment by ttrouill** [comment]
I am currently interested in a case where `beam_width > vocab_size`, and run into the following error:

```
tensorflow.python.framework.errors_impl.InvalidArgumentError: Input to reshape is a tensor with 224 values, but the requested shape has 256
	 [[Node: dynamic_seq2seq/decoder/decoder/while/BeamSearchDecoderStep/next_beam_probs/output = Reshape[T=DT_FLOAT, Tshape=DT_INT32, _device="/job:localhost/replica:0/task:0/device:CPU:0"](dynamic_seq2seq/decoder/decoder/while/BeamSearchDecoderStep/next_beam_probs/Gather, dynamic_seq2seq/decoder/decoder/while/BeamSearchDecoderStep/next_beam_probs/strided_slice)]]

Caused by op u'dynamic_seq2seq/decoder/decoder/while/BeamSearchDecoderStep/next_beam_probs/output', defined at:
  File "/home/tounze/anaconda2/lib/python2.7/runpy.py", line 174, in _run_module_as_main
    "__main__", fname, loader, pkg_name)
  File "/home/tounze/anaconda2/lib/python2.7/runpy.py", line 72, in _run_code
    exec code in run_globals
  File "/home/tounze/nmt/nmt/nmt.py", line 608, in <module>
    tf.app.run(main=main, argv=[sys.argv[0]] + unparsed)
  File "/home/tounze/anaconda2/lib/python2.7/site-packages/tensorflow/python/platform/app.py", line 124, in run
    _sys.exit(main(argv))
  File "/home/tounze/nmt/nmt/nmt.py", line 601, in main
    run_main(FLAGS, default_hparams, train_fn, inference_fn)
  File "/home/tounze/nmt/nmt/nmt.py", line 580, in run_main
    trans_file, hparams, num_workers, jobid)
  File "nmt/inference.py", line 102, in inference
    infer_model = model_helper.create_infer_model(model_creator, hparams, scope)
  File "nmt/model_helper.py", line 209, in create_infer_model
    extra_args=extra_args)
  File "nmt/attention_model.py", line 61, in __init__
    extra_args=extra_args)
  File "nmt/model.py", line 109, in __init__
    res = self.build_graph(hparams, scope=scope)
  File "nmt/model.py", line 303, in build_graph
    encoder_outputs, encoder_state, hparams)
  File "nmt/model.py", line 466, in _build_decoder
    scope=decoder_scope)
  File "/home/tounze/anaconda2/lib/python2.7/site-packages/tensorflow/contrib/seq2seq/python/ops/decoder.py", line 309, in dynamic_decode
    swap_memory=swap_memory)
  File "/home/tounze/anaconda2/lib/python2.7/site-packages/tensorflow/python/ops/control_flow_ops.py", line 2958, in while_loop
    result = loop_context.BuildLoop(cond, body, loop_vars, shape_invariants)
  File "/home/tounze/anaconda2/lib/python2.7/site-packages/tensorflow/python/ops/control_flow_ops.py", line 2722, in BuildLoop
    pred, body, original_loop_vars, loop_vars, shape_invariants)
  File "/home/tounze/anaconda2/lib/python2.7/site-packages/tensorflow/python/ops/control_flow_ops.py", line 2664, in _BuildLoop
    body_result = body(*packed_vars_for_body)
  File "/home/tounze/anaconda2/lib/python2.7/site-packages/tensorflow/contrib/seq2seq/python/ops/decoder.py", line 254, in body
    decoder_finished) = decoder.step(time, inputs, state)
  File "/home/tounze/anaconda2/lib/python2.7/site-packages/tensorflow/contrib/seq2seq/python/ops/beam_search_decoder.py", line 499, in step
    length_penalty_weight=length_penalty_weight)
  File "/home/tounze/anaconda2/lib/python2.7/site-packages/tensorflow/contrib/seq2seq/python/ops/beam_search_decoder.py", line 589, in _beam_search_step
    name="next_beam_probs")
  File "/home/tounze/anaconda2/lib/python2.7/site-packages/tensorflow/contrib/seq2seq/python/ops/beam_search_decoder.py", line 806, in _tensor_gather_helper
    output = array_ops.reshape(output, final_shape, name="output")
  File "/home/tounze/anaconda2/lib/python2.7/site-packages/tensorflow/python/ops/gen_array_ops.py", line 3998, in reshape
    "Reshape", tensor=tensor, shape=shape, name=name)
  File "/home/tounze/anaconda2/lib/python2.7/site-packages/tensorflow/python/framework/op_def_library.py", line 787, in _apply_op_helper
    op_def=op_def)
  File "/home/tounze/anaconda2/lib/python2.7/site-packages/tensorflow/python/framework/ops.py", line 3160, in create_op
    op_def=op_def)
  File "/home/tounze/anaconda2/lib/python2.7/site-packages/tensorflow/python/framework/ops.py", line 1617, in __init__
    self._traceback = self._graph._extract_stack()  # pylint: disable=protected-access

InvalidArgumentError (see above for traceback): Input to reshape is a tensor with 224 values, but the requested shape has 256
	 [[Node: dynamic_seq2seq/decoder/decoder/while/BeamSearchDecoderStep/next_beam_probs/output = Reshape[T=DT_FLOAT, Tshape=DT_INT32, _device="/job:localhost/replica:0/task:0/device:CPU:0"](dynamic_seq2seq/decoder/decoder/while/BeamSearchDecoderStep/next_beam_probs/Gather, dynamic_seq2seq/decoder/decoder/while/BeamSearchDecoderStep/next_beam_probs/strided_slice)]]
```

@shuuki4 do you have an idea?
